subprojects/core/src/main/java/org/gradle/plugin/repository/portal/CustomPluginProvider.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.plugin.repository.portal;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.Nullable;
+
+/**
+ * Allows users to add custom logic to find where a plugin is located in a repository.
+ */
+@Incubating
+public interface CustomPluginProvider {
+
+    /**
+     * Called to find a plugin implementation.
+     *
+     * @param name name of the plugin to be resolved
+     * @param version version of the plugin to be resolved against. Can be null.
+     *
+     * @return null when not found, or dependency notation in the format of a dependencyNotation as defined in {@link org.gradle.api.artifacts.dsl.DependencyHandler}
+     */
+    @Nullable Object findPlugin(String name, @Nullable String version);
```

Comment:
I'd prefer having a found/notFound callback instead of a factory plus return type.